lib/Attachment.php

```diff
@@ -0,0 +1,556 @@
+	/**
+	 * File.
+	 *
+	 * @api
+	 * @var mixed
+	 */
+	public $file;
+
+	/**
+	 * File location.
+	 *
+	 * @api
+	 * @var string The absolute path to the attachmend file in the filesystem
+	 *             (Example: `/var/www/htdocs/wp-content/uploads/2015/08/my-pic.jpg`)
+	 */
+	public $file_loc;
+
+	/**
+	 * Raw file size.
+	 *
+	 * @api
+	 * @since 2.0.0
+	 * @var int Raw file size in bytes.
+	 */
+	public $file_size_raw = null;
+
+	/**
+	 * Formatted file size.
+	 *
+	 * @api
+	 * @since 2.0.0
+	 * @var null|string File size string.
+	 */
+	public $file_size = null;
+
+	/**
+	 * File extension.
+	 *
+	 * @api
+	 * @since 2.0.0
+	 * @var null|string A file extension.
+	 */
+	public $file_extension = null;
+
+	/**
+	 * Absolute URL.
+	 *
+	 * @var string The absolute URL to the attachment.
+	 */
+	public $abs_url;
+
+	/**
+	 * Attachment ID.
+	 *
+	 * @api
+	 * @var integer The attachment ID.
+	 */
+	public $id;
+
+	/**
+	 * Attached file.
+	 *
+	 * @var array The file as stored in the WordPress database.
+	 */
+	protected $_wp_attached_file;
+
+	/**
+	 * File types.
+	 *
+	 * @var array An array of supported relative file types.
+	 */
+	private $image_file_types = array(
+		'jpg',
+		'jpeg',
+		'png',
+		'svg',
+		'bmp',
+		'ico',
+		'gif',
+		'tiff',
+		'pdf',
+	);
+
+	/**
+	 * Caption text.
+	 *
+	 * @api
+	 * @var string The caption that is stored as post_excerpt in the posts table in the database.
+	 */
+	public $caption;
+
+	/**
+	 * Creates a new `Timber\Attachment` object.
+	 *
+	 * @api
+	 * @example
+	 * ```php
+	 * // You can pass it an ID number
+	 * $myImage = new Timber\Attachment(552);
+	 *
+	 * // Or send it a URL to an image
+	 * $myImage = new Timber\Attachment( 'http://google.com/logo.jpg' );
+	 * ```
+	 *
+	 * @param int|mixed $attachment An attachment ID, a `Timber\Post`, a `WP_Post` object, an ACF
+	 *                              image array, a path (absolute or relative) or an URL.
+	 */
+	public function __construct( $attachment ) {
+		$this->init( $attachment );
+	}
+
+	/**
+	 * Gets the src for an attachment.
+	 *
+	 * @api
+	 *
+	 * @return string The src of the attachment.
+	 */
+	public function __toString() {
+		return $this->src();
+	}
+
+	/**
+	 * Inits the object.
+	 *
+	 * @internal
+	 *
+	 * @param int|mixed $iid An attachment identifier.
+	 */
+	public function init( $iid = null ) {
+		// Make sure we actually have something to work with.
+		if ( ! $iid ) {
+			Helper::error_log( 'Initialized Timber\Attachment without providing first parameter.' );
+
+			return null;
+		}
+
+		/**
+		 * If passed a Timber\Attachment object, grab the ID and continue. Otherwise, try to check
+		 * for an ACF image array an take the ID from that array.
+		 */
+		if ( $iid instanceof Attachment ) {
+			$iid = (int) $iid->ID;
+		} elseif ( is_array( $iid ) && isset( $iid['ID'] ) ) {
+			// Assume ACF image array.
+			$iid = $iid['ID'];
+		}
+
+		if ( ! is_numeric( $iid ) && is_string( $iid ) ) {
+			if ( strstr( $iid, '://' ) ) {
```

Comment:
I think this should go in another class. We mix WordPress attachments with external images here. I think of another `ExternalImage` class which does not extend `Image` as they don't share the same properties. Any thoughts? @gchtr @jarednova

Reply:
@pascalknecht @jarednova For me, it would make sense that an image that is loaded through an URL or a path wouldn’t have the same methods and properties as an image that is an actual attachment. I like the concept, but I’m not yet sure about the name «External», maybe `ImageAsset`? And we couldn’t return a new class here, right? Is this where object factories would come in handy?

Reply:
@gchtr I would probably implement a new class. The advantage of object factories is mostly that you can decouble the instantiation of objects from the classes so no real benefit here I think. I would suggest implementing a new class `ImageAsset` or `ExternalImage` which doesn't have all the empty properties it currently has.